x-pack/plugins/security_solution/public/flyout/url/use_sync_flyout_state_with_url.tsx

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useRef } from 'react';
+import type { ExpandableFlyoutApi, ExpandableFlyoutContext } from '@kbn/expandable-flyout';
+import { useSyncToUrl } from '@kbn/securitysolution-url-hooks';
+import last from 'lodash/last';
+
+const URL_KEY = 'alert-flyout' as const;
+
+type FlyoutState = Parameters<ExpandableFlyoutApi['openFlyout']>[0];
+
+/**
+ * Sync flyout state with the url and open it when relevant url state is detected in the query string
+ * @returns [ref, flyoutChangesHandler]
+ */
+export const useSyncFlyoutStateWithUrl = () => {
+  const flyoutApi = useRef<ExpandableFlyoutApi>(null);
+
+  const syncStateToUrl = useSyncToUrl<FlyoutState>(URL_KEY, (data) => {
+    flyoutApi.current?.openFlyout(data);
+  });
+
+  // This should be bound to flyout changed and closed events.
+  // When flyout is closed, url state is cleared
+  const handleFlyoutChanges = useCallback(
+    (state?: ExpandableFlyoutContext['panels']) => {
+      if (!state) {
+        return syncStateToUrl(undefined);
+      }
+
+      return syncStateToUrl({
+        ...state,
+        preview: last(state.preview),
+      });
+    },
+    [syncStateToUrl]
+  );
+
+  return [flyoutApi, handleFlyoutChanges] as const;
```

Comment:
this is more a question for my own knowledge, why returning an array instead of an object here? I've encountered other places where we return arrays and I don't find it so nice to use

Reply:
with this approach, you are free to name the handler however you want a bit shorter than you would with the object. just like you an do it with setter and value returned from `useState()`